Write SOP (resolves #85)

jvivian opened the pull request.

jvivian committed:
Write SOP (resolves #85)

jvivian commented:
@hannes-ucsc — This is just a first stab. I figured I'd let @JakeNarkizian use this as his template for the next production run and he can give me feedback if anything is unclear.

hannes-ucsc commented on Production_SOP.md:
```diff
@@ -0,0 +1,71 @@
+3. `pip install cgcloud-toil==1.5.5`
+4. `deactivate`
+5. `ln -snf ~/cgcloud/bin/activate ~/bin/`
+6. Set `~/.profile` options from CGCloud README
```
Be explicit. No linking or references except for optional steps. This document should be self contained. Be sure to mention `CGCLOUD_KEYPAIRS`.

hannes-ucsc replied:
```
The following steps are provided for referecnce but it is recommended to read the source documentation to gain full understanding on how to use CGCloud.
```

jvivian replied:
Substituted in your comment, added `~/.profile` options.

hannes-ucsc commented on Production_SOP.md:
```diff
@@ -0,0 +1,71 @@
+ 
+## Environment Setup
+ 
+ These steps should only have to be run once. Toil-box images will have to be updated periodically to stay current.
```
So for readability you should always backquote strings that a user would type into a command line.

So for example:

```
Bash can be started by running the `bash` command.
```

Or in this case:

```
The `toil-box` image will …
```

hannes-ucsc replied:
"periodically" is imprecise. When exactly should they update?

jvivian replied:
> "periodically" is imprecise. When exactly should they update?

That's a good question, I replaced it with: 

" `toil-box` images will have to be updated following major releases of Toil that contain critical features or bug fixes, or at the direction of the Toil development team."

hannes-ucsc commented on Production_SOP.md:
```diff
@@ -0,0 +1,71 @@
+ These steps should only have to be run once. Toil-box images will have to be updated periodically to stay current.
+  
+### Setup CGCloud
+Comprehensive help is found at: https://github.com/BD2KGenomics/cgcloud .
```
Markdown can do hyperlinks.

jvivian replied:
Corrected.

hannes-ucsc commented on Production_SOP.md:
```diff
@@ -0,0 +1,71 @@
+### Setup and Run Pipeline
+SSH into leader, install toil-scripts, setup config and manifest, grow cluster, and start run.
+
+1. `cgcloud ssh -c example toil-leader`
```
Mention briefly that `-c` is same as `--cluster-name`

jvivian replied:
I replaced with the full name for clarity.

hannes-ucsc commented on Production_SOP.md:
```diff
@@ -0,0 +1,71 @@
+8. `cgcloud grow-cluster --cluster-name example --spot-auto-zone --spot-bid 1.00 --instance-type c3.8xlarge --num-workers 22 toil`
+9. `cgcloud ssh -c example toil-leader`
+10. Place boto or AWS credentials on the leader (`~/.boto` or `~/.aws/credentials`)
+10. `screen -S example-run`
```
10 is duplicated

jvivian replied:
Corrected.

hannes-ucsc commented on Production_SOP.md:
```diff
@@ -0,0 +1,71 @@
+2. `virtualenv --system-site-packages toilscripts`
+3. `source toilscripts/bin/activate`
+4. `pip install toil-scripts==2.0.11`
+5. Run the binary for the pipeline, example: `toil-rnaseq` for the RNA-seq CGL pipeline.
```
`toil-rnaseq`  is not a binary

hannes-ucsc replied:
Isn't it "RNASeq"?

jvivian replied:
/s/binary/entrypoint.

Typically I see [RNA-seq](https://en.wikipedia.org/wiki/RNA-Seq)

hannes-ucsc commented on Production_SOP.md:
```diff
@@ -0,0 +1,71 @@
+5. Run the binary for the pipeline, example: `toil-rnaseq` for the RNA-seq CGL pipeline.
+6. Generate a config and manifest file, e.g. `toil-rnaseq generate`. These files will be generated in the current working directory, and contain instructions on how to fill them in.
+7. `exit` so we can grow the cluster, in this case by 22 workers, ideally using spot market.
+8. `cgcloud grow-cluster --cluster-name example --spot-auto-zone --spot-bid 1.00 --instance-type c3.8xlarge --num-workers 22 toil`
```
Add section on choosing spot market price.

jvivian replied:
Added section.

hannes-ucsc commented on Production_SOP.md:
```diff
@@ -0,0 +1,71 @@
+## Handling Errors
+
+Sometimes a pipeline fails. This can be for various reasons, but it is unlikely that there is an issue with the pipeline
+or Toil itself. Failures generally stem from an issue with the sample -- improperly paired reads, samples too large for
```
I like your confidence but there is little evidence for the claim in the second sentence.

jvivian replied:
I removed this section completely.

hannes-ucsc commented on Production_SOP.md:
```diff
@@ -0,0 +1,71 @@
+Sometimes a pipeline fails. This can be for various reasons, but it is unlikely that there is an issue with the pipeline
+or Toil itself. Failures generally stem from an issue with the sample -- improperly paired reads, samples too large for
+the amount of memory or disk on a machine, etc.
+This assumption isn't based on hubris, but the fact that the pipeline had a throughput of 99.6% across 20,000 samples.
```
This sounds arrogant. It was a different version of the pipeline and a different version of Toil. And we had plenty of problems during the run. The fact that Toil+Pipeline were able to run 20k samples with plenty of hand-holding doesn't mean that Joe Shmo should have no problems with their 20 samples.

jvivian replied:
I agree, and not appropriate for technical documentation. I just removed this section.

jvivian committed:
Reviewer corrections

Added section on spot market pricing
Removed unnecessary statements in Handling Errors section
Added CGCloud profile options

jvivian commented:
@hannes-ucsc — Ready for review. I believe @JakeNarkizian plans on running the next production batch, so maybe he can use this SOP for that and provide feedback first.

nsjake commented:
@jvivian I started a production run on monday night already, but I can simulate a production order.
